### PR TITLE
[CVP][LVI] Add support for InsertElementInst in LVI

### PR DESCRIPTION
**Pull request by rajatbajpai**
Currently, the LVI analysis pass doesn't support InsertElementInst vector instruction. Due to this, some optimization opportunities are missed. For example, in the below example, ICMP instruction can be folded but it doesn't.

```
...
%ie1 = insertelement <2 x i32> poison, i32 10, i64 0
%ie2 = insertelement <2 x i32> %ie1, i32 20, i64 1
%icmp = icmp <2 x i1> %ie2, <i32 40, i32 40>
...
```

This change adds InsertElementInst support in the LVI analysis pass to fix the motivating example.

### CODE REVIEW

File: llvm/lib/Analysis/LazyValueInfo.cpp
Comment (rajatbajpai) [reply llvm/lib/Analysis/LazyValueInfo.cpp:1074]: I tried `getBlockValue` initially but test2 fails without `ConstantVector` logic. Maybe `getBlockValue` needs handling for `ConstantVector`?

Comment (nikic) [reply llvm/lib/Analysis/LazyValueInfo.cpp:1074]: I think this is due to a weakness in mergeIn(). I'll fix it.

File: llvm/test/Transforms/CorrelatedValuePropagation/insertelement.ll
Comment (rajatbajpai) [reply llvm/test/Transforms/CorrelatedValuePropagation/insertelement.ll]: Fixed in the latest revision.

File: llvm/lib/Analysis/LazyValueInfo.cpp
Comment (nikic) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: https://github.com/llvm/llvm-project/pull/99466 is the fix I had in mind.

Comment (rajatbajpai) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: Ah I see, I'll wait for that PR to be merged then.

Comment (nikic) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: I've merged the change, please give it another try.

Comment (rajatbajpai) [reply llvm/lib/Analysis/LazyValueInfo.cpp]: Okay, it seems to be working. I have updated the review. Thanks!

**Review by nikic — APPROVED** [review]
LGTM

**Comment by rajatbajpai** [comment]
Will merge the changes after getting the clean pipeline. Thank you for reviewing this change.

**Comment by rajatbajpai** [comment]
The Linux build is failing and the error doesn't seem to be related to my change. I'll wait some time for the setup to recover.

**Comment by nikic** [comment]
@rajatbajpai You can ignore the linux build.

**Comment by rajatbajpai** [comment]
@nikic could you please merge this PR for me? I don't have permission to merge a PR. Thanks!

**Comment by mikaelholmen** [comment]
Hello,

I bisected a crash back to this patch:
```opt -passes="correlated-propagation" bbi-101893.ll -o /dev/null -S```
crashes with
```
opt: ../include/llvm/IR/Instructions.h:2647: void llvm::PHINode::setIncomingValue(unsigned int, Value *): Assertion `getType() == V->getType() && "All operands to PHI node must be the same type as the PHI node!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build-all/bin/opt -passes=correlated-propagation bbi-101893.ll -o /dev/null -S
1.	Running pass "function(correlated-propagation)" on module "bbi-101893.ll"
2.	Running pass "correlated-propagation" on function "func_18"
 #0 0x0000562b7e9082a6 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x444a2a6)
 #1 0x0000562b7e905cce llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x4447cce)
 #2 0x0000562b7e908b59 SignalHandler(int) Signals.cpp:0:0
 #3 0x00007fa6b79bbcf0 __restore_rt (/lib64/libpthread.so.0+0x12cf0)
 #4 0x00007fa6b5574acf raise (/lib64/libc.so.6+0x4eacf)
 #5 0x00007fa6b5547ea5 abort (/lib64/libc.so.6+0x21ea5)
 #6 0x00007fa6b5547d79 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d79)
 #7 0x00007fa6b556d426 (/lib64/libc.so.6+0x47426)
 #8 0x0000562b8007d234 runImpl(llvm::Function&, llvm::LazyValueInfo*, llvm::DominatorTree*, llvm::SimplifyQuery const&) CorrelatedValuePropagation.cpp:0:0
 #9 0x0000562b80077e54 llvm::CorrelatedValuePropagationPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x5bb9e54)
#10 0x0000562b7fd25b5d llvm::detail::PassModel<llvm::Function, llvm::CorrelatedValuePropagationPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#11 0x0000562b7eb25057 llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x4667057)
#12 0x0000562b7fd28a6d llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#13 0x0000562b7eb29cbb llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x466bcbb)
#14 0x0000562b7fd21fdd llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilderPipelines.cpp:0:0
#15 0x0000562b7eb23ce7 llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x4665ce7)
#16 0x0000562b7fcc211c llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x580411c)
#17 0x0000562b7e8cb60a optMain (build-all/bin/opt+0x440d60a)
#18 0x00007fa6b5560d85 __libc_start_main (/lib64/libc.so.6+0x3ad85)
#19 0x0000562b7e8c91ee _start (build-all/bin/opt+0x440b1ee)
Abort (core dumped)
```[bbi-101893.ll.gz](https://github.com/user-attachments/files/17974801/bbi-101893.ll.gz)

**Comment by nikic** [comment]
Reduced test case:
```
@g = external global i32

define <2 x i16> @test() {
  %ins = insertelement <2 x i16> poison, i16 ptrtoint (ptr @g to i16), i32 0
  ret <2 x i16> %ins
}
```

**Comment by nikic** [comment]
@mikaelholmen Should be fixed by https://github.com/llvm/llvm-project/commit/7a7a426188eea0b181738436205759e13ad6fd7b.

**Comment by mikaelholmen** [comment]
That was quick. Thanks!

**Pull request merged**
